cpp/src/arrow/extension_type.h

```diff
@@ -0,0 +1,103 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+/// User-defined extension types. EXPERIMENTAL in 0.13.0
+/// \since 0.13.0
+
+#pragma once
+
+#include <memory>
+#include <string>
+
+#include "arrow/array.h"
+#include "arrow/type.h"
+#include "arrow/util/visibility.h"
+
+namespace arrow {
+
+class TypeVisitor;
+
+class ExtensionTypeTraits {
+ public:
+  virtual ~ExtensionTypeTraits() = default;
+
+  /// \brief Name used for serialization, and lookups for deeserialization
+  virtual std::string extension_name() const = 0;
+
+  /// \brief Determine if two instances of the same extension types are
+  /// equal. Invoked from ExtensionType::Equals
+  /// \param[in] other the type to compare this type with
+  /// \return bool true if type instances are equal
+  virtual bool ExtensionEquals(const ExtensionType& other) const = 0;
+};
+
+/// \brief The base class for custom / user-defined types.
+class ARROW_EXPORT ExtensionType : public DataType, public ExtensionTypeTraits {
```

Comment:
I'm a bit wary about multiple inheritance here.
How about instead store a `shared_ptr<ExtensionTypeAdapter>` in the `ExtensionType` and use that to query extension name etc.

Reply:
That's fine, will do

Reply:
So currently `ExtensionTypeTraits` is a mechanism to ensure that `ExtensionType` remains an abstract type, requiring the implementation of its methods.

If an instance of ExtensionType needs to have an `ExtensionTypeAdapter`, where does that come from? I do not think that each instance of `ExtensionType` should call `std::make_shared<MyExtensionTypeAdapter>()` or similar.

Reply:
Well, if the concern is the `shared_ptr` instantiation cost, the singleton pattern can be used.
Also, why can't the abstract methods be defined directly on `ExtensionType`?